File: src/main/java/com/refinedmods/refinedstorage/screen/ExporterScreen.java
package com.refinedmods.refinedstorage.screen;

import com.refinedmods.refinedstorage.RS;
import com.refinedmods.refinedstorage.blockentity.ExporterBlockEntity;
import com.refinedmods.refinedstorage.blockentity.NetworkNodeBlockEntity;
import com.refinedmods.refinedstorage.container.ExporterContainerMenu;
import com.refinedmods.refinedstorage.item.UpgradeItem;
import com.refinedmods.refinedstorage.screen.widget.sidebutton.ExactModeSideButton;
import com.refinedmods.refinedstorage.screen.widget.sidebutton.RedstoneModeSideButton;
import com.refinedmods.refinedstorage.screen.widget.sidebutton.TypeSideButton;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Inventory;

public class ExporterScreen extends BaseScreen<ExporterContainerMenu> {
    private static final ResourceLocation TEXTURE = new ResourceLocation(RS.ID, "textures/gui/exporter.png");

    private boolean hasRegulatorMode;

    public ExporterScreen(ExporterContainerMenu containerMenu, Inventory playerInventory, Component title) {
        super(containerMenu, 211, 137, playerInventory, title);

        this.hasRegulatorMode = hasRegulatorMode();
    }

    private boolean hasRegulatorMode() {
        return ((ExporterBlockEntity) menu.getBlockEntity()).getNode().getUpgrades().hasUpgrade(UpgradeItem.Type.REGULATOR);
    }

    @Override
    public void onPostInit(int x, int y) {
        addSideButton(new RedstoneModeSideButton(this, NetworkNodeBlockEntity.REDSTONE_MODE));

        addSideButton(new TypeSideButton(this, ExporterBlockEntity.TYPE));

        addSideButton(new ExactModeSideButton(this, ExporterBlockEntity.COMPARE));
    }

    @Override
    public void tick(int x, int y) {
        boolean updatedHasRegulatorMode = hasRegulatorMode();
        if (hasRegulatorMode != updatedHasRegulatorMode) {
            hasRegulatorMode = updatedHasRegulatorMode;

            menu.initSlots();
        }
    }

    @Override
    public void renderBackground(GuiGraphics graphics, int x, int y, int mouseX, int mouseY) {
        graphics.blit(TEXTURE, x, y, 0, 0, imageWidth, imageHeight);
    }

    @Override
    public void renderForeground(GuiGraphics graphics, int mouseX, int mouseY) {
        renderString(graphics, 7, 7, title.getString());
        renderString(graphics, 7, 43, I18n.get("container.inventory"));
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/screen/ExporterScreen.kt
package com.refinedmods.refinedstorage.screen

import com.mojang.blaze3d.matrix.MatrixStack
import com.refinedmods.refinedstorage.RS
import com.refinedmods.refinedstorage.container.ExporterContainer
import com.refinedmods.refinedstorage.item.UpgradeItem
import com.refinedmods.refinedstorage.screen.widget.sidebutton.ExactModeSideButton
import com.refinedmods.refinedstorage.screen.widget.sidebutton.RedstoneModeSideButton
import com.refinedmods.refinedstorage.screen.widget.sidebutton.TypeSideButton
import com.refinedmods.refinedstorage.tile.ExporterTile
import net.minecraft.client.resources.I18n
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.util.text.Text

class ExporterScreen(container: ExporterContainer, playerInventory: PlayerInventory?, title: Text?) : BaseScreen<ExporterContainer?>(container, 211, 137, playerInventory, title) {
    private var hasRegulatorMode: Boolean
    private fun hasRegulatorMode(): Boolean {
        return (container.getTile() as ExporterTile).getNode().getUpgrades().hasUpgrade(UpgradeItem.Type.REGULATOR)
    }

    override fun onPostInit(x: Int, y: Int) {
        addSideButton(RedstoneModeSideButton(this, ExporterTile.REDSTONE_MODE))
        addSideButton(TypeSideButton(this, ExporterTile.TYPE))
        addSideButton(ExactModeSideButton(this, ExporterTile.COMPARE))
    }

    override fun tick(x: Int, y: Int) {
        val updatedHasRegulatorMode = hasRegulatorMode()
        if (hasRegulatorMode != updatedHasRegulatorMode) {
            hasRegulatorMode = updatedHasRegulatorMode
            container.initSlots()
        }
    }

    override fun renderBackground(matrixStack: MatrixStack?, x: Int, y: Int, mouseX: Int, mouseY: Int) {
        bindTexture(RS.ID, "gui/exporter.png")
        blit(matrixStack, x, y, 0, 0, xSize, ySize)
    }

    override fun renderForeground(matrixStack: MatrixStack?, mouseX: Int, mouseY: Int) {
        renderString(matrixStack, 7, 7, title.getString())
        renderString(matrixStack, 7, 43, I18n.format("container.inventory"))
    }

    init {
        hasRegulatorMode = hasRegulatorMode()
    }
}